### PR TITLE
Einsum GAT

### PR DESCRIPTION
**Pull request by cgarciae**
### Status
As of now this is just an initial port of the Multi Head Attention code I was using. I temporary put it in its own file to make initial development easier. 

### Objective 
The main objective will be to modify the `GraphAttention` class and `multi_head_attention` function (maybe merge them) as to be compatible with the current implementation.

BTW: I might be a good to eventually break up the `convolutional` module into a folder with each layer in its own file, it would make contribution easier.

### CODE REVIEW

**Comment by cgarciae** [comment]
@danielegrattarola Where does this equivalence come from?
```
[[a_1], [a_2]]^T [[Wh_i], [Wh_2]] = [a_1]^T [Wh_i] + [a_2]^T [Wh_j]
```
Addition must be broadcasted in both dimensions for this to make sense right?

**Comment by cgarciae** [comment]
BTW: Now that I am implementing this I this I see that the GAT layer is very peculiar. I am just wondering if using the more standard MultiHead Attention layer from Transformers masked by the adjacency matrix would give similar results.

**Comment by cgarciae** [comment]
@danielegrattarola I got a working implementation and tested it with `node_classification_gat.py` example. Notes:

- Trains!
- Code looks clean
- Slightly slower :(
  - Branched: `50-60 us` per batch
  - Einsum: `60-70 us` per batch

Don't know if something in my code is causing the slowdown. Might be that `matmul` is faster than `einsum` for this case.

**Comment by danielegrattarola** [comment]
Hi,
sorry but I've been super busy recently and I haven't had a chance to look at the code. I'll have a look at it ASAP and see how to integrate it in the library. 
The code seems good, I'm not sure why einsum is slowing things down (it shouldn't at all). 

Cheers, 
D

**Comment by danielegrattarola** [comment]
So I've checked the code and everything seems ok. The new layer is slower to train but it's significantly faster in the forward pass. 
Overall, it seems to me that the new layer is more readable/clean, so I'm ready to merge into the main convolution file. 

I've made a mess with the commits and added a bunch of stuff from the latest develop branch, but you can review the changes to your file.

**Comment by cgarciae** [comment]
@danielegrattarola I'll check it out. I didn't want to add support for the other arguments until you gave a look a the implementation, thanks for finishing the job!

**Comment by cgarciae** [comment]
@danielegrattarola replaced `GraphAttention` with the new implementation in `covolutional.py`. This PR is good to go.

**Comment by danielegrattarola** [comment]
Looks good, thanks for making this!!

Cheers, 
D

**Comment by cgarciae** [comment]
@danielegrattarola Just noticed that the `activity_regularizer` parameter is not being used, should we use it or remove it?

**Comment by danielegrattarola** [comment]
Hi, 
even if it's not explicitly used in the layer, it should still be picked up by Keras. If you backtrace the various calls to `super().init(**kwargs)` you'll get to `Layer`: the activity regularizer is used on [this line](https://github.com/tensorflow/tensorflow/blob/b2ea4cc49d7ff501ec7bdbfcee7bfdaa3e3befd2/tensorflow/python/keras/engine/base_layer.py#L315).

I could be mistaken, but I think it has always worked like this even in stand-alone Keras.

Let me know what you think. 
Cheers, 
D

**Comment by cgarciae** [comment]
I see, then we need to remove [this line](https://github.com/cgarciae/spektral/blob/d14a5f5c37c7685fb4b1bad491bb7b2d6ccbf4b7/spektral/layers/convolutional.py#L857) so it goes into the `**kwargs` instead.

**Comment by danielegrattarola** [comment]
You're absolutely right, I don't know what I was thinking. 
I need to fix all other layers as well, I think got misguided by [the Dense layer in stand-alone Keras](https://github.com/keras-team/keras/blob/7a39b6c62d43c25472b2c2476bd2a8983ae4f682/keras/layers/core.py#L859-L873).

**Comment by danielegrattarola** [comment]
@cgarciae should have fixed the issue for all layers in commit 2b92e8b816d2b6b093f3bda07dcf8dc2ba973297. Thanks for noticing!!
